_posts/2019-03-19-til-validations-and-constraints.md

```diff
@@ -0,0 +1,229 @@
+> - [Wikipedia](https://en.wikipedia.org/wiki/Data_integrity)
+
+So we're building a super cool app with users that can login and logout.
+We'd probably have some schema like this:
+```elixir
+# Using Phoenix 1.4 with Contexts but still applies all the same
+defmodule MyCoolWebApp.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :display_name, :string
+    field :email, :string
+    field :password_hash, :string
+    field :password, :string, virtual: true
+
+    timestamps()
+  end
+end
+```
+
+And inside of it, we'd want to describe the `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+end
+```
+
+If we only had this, we might have a few headaches come our way. It is very easy to
+prematurely submit a form without filling all the fields out. Potentially, now the
+user's profile has a lovely email or password of `nil`, or worse `""`, in the
+database. This would suck, for the user(s) and the developer(s).
+
+So in order to fix, we'd use a validation!
+
+### Validations
+Many of the validations we have in Ecto will be executed without needing
+to interact with the database. That means the validation will be executed prior to the
+attempt of inserting or updating something in the database. If we wanted to
+insert a new user into our database, we'd first want to make sure there is data
+inside of the changeset.
+
+Let's add the [`validate_require/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#validate_required/3) into our `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+  |> validate_required([:display_name, :email, :password])
+end
+```
+
+And for free, we get a set of errors coming out if a user were to make a mistake:
+```elixir
+iex(1)> %User{} |> User.changeset(%{})
+#Ecto.Changeset<
+  action: nil,
+  changes: %{},
+  errors: [
+    display_name: {"can't be blank", [validation: :required]},
+    email: {"can't be blank", [validation: :required]},
+    password: {"can't be blank", [validation: :required]}
+  ],
+  data: #MyCoolWebApp.Accounts.User<>,
+  valid?: false
+>
+```
+
+There are a ton of validations out there that can enhance your app! Take a
+look at the documentation for the [Ecto.Changeset](https://hexdocs.pm/ecto/Ecto.Changeset.html#summary). Now the next
+phase will be looking at when we need to apply some constraints.
+
+### Constraints
+You might have gotten to this point and thought, why are there differences?
+Let's think about many apps that we use. When we sign up for an application, are
+we allowed to signup with the same email as another user? (Hint: the answer should
+always be no.) The reason is because developers want to be able to distinguish users
+between one another and ensure that they can send the right person the right information.
+
+So why couldn't we just have a validation for uniqueness? Well remember the definition
```

Comment:
Instead of "Well remember the definition we learned about validtaions" -> "Remember that, by definition validations are executed *prior to checking the database*"